src/adagrams.js

```diff
@@ -1,15 +1,187 @@
+  }
+
+  // use the random indices to get letters
+  const lettersInHand = [];
+  for (let index of randomIndices) {
+    lettersInHand.push(weightedPool[index]);
+  }
+
+  return lettersInHand;
 };
 
+////////// WAVE 2 //////////
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+  // change input from string to array
+  const lettersPlayed = input.split("");
+
+  // make a copy of our hand
+  const lettersInHandCopy = [...lettersInHand];
+
+  // if letter is in our hand, remove it. if it isn't, return false
+  for (let letter of lettersPlayed) {
+    if (lettersInHandCopy.includes(letter)) {
+      // remove one letter from our hand, then break out
+      for (let i = 0; i < lettersInHandCopy.length; i++) {
+        if (lettersInHandCopy[i] === letter) {
+          lettersInHandCopy.splice(i, 1);
+          break;
+        }
+      }
+    } else {
+      return false;
+    }
+  }
+
+  // if all letters were in our hand, return true
+  return true;
 };
 
+////////// WAVE 3 //////////
 export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  // change input from string to array
+  const wordLetters = word.split("");
+
+  // get the score of each letter and add to wordScore
+  let wordScore = 0;
+  for (let letter of wordLetters) {
+    // get letter score from upper- or lowercase letters
+    let letterScore = LETTER_SCORES[letter.toUpperCase()];
+    wordScore = wordScore + letterScore;
+  }
+
+  // add length bonus to wordScore
+  if (wordLetters.length >= 7) {
+    wordScore = wordScore + 8;
+  }
+
+  return wordScore;
 };
 
+////////// WAVE 4 ///////////
 export const highestScoreFrom = (words) => {
-  // Implement this method for wave 4
+  // find highest score
+  let highScore = 0;
+  for (let word of words) {
+    // call function from wave 3
+    let wordScore = scoreWord(word);
+    if (wordScore > highScore) {
+      highScore = wordScore;
+    }
+  }
+
+  // make a list of all words w/ highest score
+  let highestScoringWords = [];
+  for (let word of words) {
+    let wordScore = scoreWord(word);
+    if (wordScore === highScore) {
+      highestScoringWords.push(word);
+    }
+  }
+
+  // if highestScoringWords has only one word, return the word
+  if (highestScoringWords.length === 1) {
+    return { word: highestScoringWords[0], score: highScore };
+  }
+
+  // if a word in highestScoringWords has 10 letters, return the word
+  for (let word of highestScoringWords) {
+    if (word.length === 10) {
+      return { word: word, score: highScore };
+    }
+  }
+
+  // find length of the shortest word
+  let shortestLength = 100;
+  for (let word of highestScoringWords) {
+    if (word.length < shortestLength) {
+      shortestLength = word.length;
+    }
+  }
+
+  // find shortest word and return it
+  for (let word of highestScoringWords) {
+    if (word.length === shortestLength) {
+      return { word: word, score: highScore };
+    }
+  }
```

Comment:
Rather than creating a variable to hold the `shortestLength`, consider checking if the current word is shorter than the current highest scoring word 👍🏾